<launch>

    <arg name="map_size_x" default="10.0"/>
    <arg name="map_size_y" default="10.0"/>
    <arg name="map_size_z" default=" 2.0"/>

    <arg name="start_x" default=" 0.0"/>
    <arg name="start_y" default=" 0.0"/>
    <arg name="start_z" default=" 0.0"/>

    <arg name="start_vx" default=" 0.2"/>
    <arg name="start_vy" default=" 0.2"/>
    <arg name="start_vz" default=" 0.0"/>
    <arg name="map_resolution" default=" 0.2"/>

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find motion_planner)/launch/rviz_config/motion_planner.rviz"
          required="true"/>

    <node pkg="motion_planner" type="run_planner" name="run_planner" output="screen" required="true">
        <remap from="~waypoints" to="/waypoint_generator/waypoints"/>
        <remap from="~map" to="/random_complex/global_map"/>

        <param name="map/margin" value="0.0"/>
        <param name="map/resolution" value="$(arg map_resolution)"/>
        <param name="map/x_size" value="$(arg map_size_x)"/>
        <param name="map/y_size" value="$(arg map_size_y)"/>
        <param name="map/z_size" value="$(arg map_size_z)"/>

        <param name="planning/start_x" value="$(arg start_x)"/>
        <param name="planning/start_y" value="$(arg start_y)"/>
        <param name="planning/start_z" value="$(arg start_z)"/>
        <param name="planning/start_vx" value="$(arg start_vx)"/>
        <param name="planning/start_vy" value="$(arg start_vy)"/>
        <param name="planning/start_vz" value="$(arg start_vz)"/>


        <param name="kinematic_astar/steering_angle" value="15.0"/>
        <param name="kinematic_astar/steering_angle_discrete_num" value="1"/>
        <param name="kinematic_astar/wheel_base" value="2.0"/>
        <param name="kinematic_astar/segment_length" value="1.6"/>
        <param name="kinematic_astar/segment_length_discrete_num" value="8"/>
        <param name="kinematic_astar/steering_penalty" value="1.5"/>
        <param name="kinematic_astar/reversing_penalty" value="3.0"/>
        <param name="kinematic_astar/steering_change_penalty" value="2.0"/>
        <param name="kinematic_astar/shot_distance" value="5.0"/>
        <param name="kinematic_astar/track" value="1.3"/>
        <param name="kinematic_astar/vehicle_width" value="0.5"/>
        <param name="kinematic_astar/vehicle_length" value="1.0"/>
        <param name="kinematic_astar/grid_size_phi" value="72"/>

    </node>

    <node pkg="motion_planner" name="random_complex" type="random_complex" output="screen">

        <param name="init_state_x" value="$(arg start_x)"/>
        <param name="init_state_y" value="$(arg start_y)"/>

        <param name="map/x_size" value="$(arg map_size_x)"/>
        <param name="map/y_size" value="$(arg map_size_y)"/>
        <param name="map/z_size" value="$(arg map_size_z)"/>

        <param name="map/circle_num" value="40"/>
        <param name="map/obs_num" value="300"/>
        <param name="map/resolution" value="0.1"/>

        <param name="ObstacleShape/lower_rad" value="0.1"/>
        <param name="ObstacleShape/upper_rad" value="0.7"/>
        <param name="ObstacleShape/lower_hei" value="1.0"/>
        <param name="ObstacleShape/upper_hei" value="3.0"/>

        <param name="CircleShape/lower_circle_rad" value="0.6"/>
        <param name="CircleShape/upper_circle_rad" value="2.0"/>

        <param name="sensing/rate" value="0.5"/>
    </node>

    <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
        <remap from="~goal" to="/goal"/>
        <param name="waypoint_type" value="manual-lonely-waypoint"/>
    </node>

    <node pkg="octomap_server" type="octomap_server_node" name="octomap_server">
                <param name="resolution" value="$(arg map_resolution)" />
                <param name="frame_id" type="string" value="world" />
                <param name="sensor_model/hit" value="0.65" />
                <param name="sensor_model/miss" value="0.1" />
                <param name="sensor_model/max_range" value="20.0" />
                <param name="occupancy_min_z" value="-2.0" />
                <param name="occupancy_max_z" value="15.0" />
                <param name="pointcloud_min_z" value="-2.0" />
                <param name="pointcloud_max_z" value="15.0" />
                <remap from="cloud_in" to="/run_planner/grid_map_vis" /> 
  </node>

</launch>